src/main/java/org/opensearch/ad/model/Entity.java

```diff
@@ -99,28 +179,205 @@ public boolean equals(Object o) {
+        return Joiner.on(",").withKeyValueSeparator("=").join(attributes);
+    }
+
+    /**
+    * Create model Id out of detector Id and attribute name and value pairs
+    *
+    * HCAD v1 uses the categorical value as part of the model document Id, but
+    *  OpenSearch’s document Id can be at most 512 bytes. Categorical values are
+    *  usually less than 256 characters, but can grow to 32766 in theory.
+    *  HCAD v1 skips an entity if the entity's name is more than 256 characters.
+    *  We cannot do that in v2 as that can reject a lot of entities. To overcome
+    *  the obstacle, we hash categorical values to a 128-bit string (like SHA-1
+    *  that git uses) and use the hash as part of the model document Id.
+    *
+    * We have choices to make regarding when to use the hash as part of a model
+    * document Id: for all HC detectors or a HC detector with multiple categorical
+    * fields. The challenge lies in providing backward compatibility of looking for
+    * a model checkpoint in the case of a HC detector with one categorical field.
+    * If using hashes for all HC detectors, we need two get requests to ensure that
+    * a model checkpoint exists. One uses the document Id without a hash, while one
+    * uses the document Id with a hash. The dual get requests are ineffective. If
+    * limiting hashes to a HC detector with multiple categorical fields, there is
+    * no backward compatibility issue. However, the code will be branchy. One may
+    * wonder if backward compatibility can be ignored; indeed, the old checkpoints
+    * will be gone after a transition period during upgrading. During the transition
+    * period, HC detectors can experience unnecessary cold starts as if the
+    * detectors were just started. Checkpoint index size can double if every model
+    * has two model documents. The transition period can be as long as 3 days since
+    * our checkpoint retention period is 3 days. There is no perfect solution. We
+    * prefer limiting hashes to an HC detector with multiple categorical fields as
+    * its customer impact is none.
+    *
+    * @param detectorId Detector Id
+    * @param attributes Attributes of an entity
+    * @return the model Id
+    */
+    public static Optional<String> getModelId(String detectorId, SortedMap<String, String> attributes) {
+        if (attributes.isEmpty()) {
+            return Optional.empty();
+        } else if (attributes.size() == 1) {
+            for (Map.Entry<String, String> categoryValuePair : attributes.entrySet()) {
+                // For OpenSearch, the limit of the document ID is 512 bytes.
+                // skip an entity if the entity's name is more than 256 characters
+                // since we are using it as part of document id.
+                String categoricalValue = categoryValuePair.getValue().toString();
+                if (categoricalValue.length() > AnomalyDetectorSettings.MAX_ENTITY_LENGTH) {
```

Comment:
What is the consequence for returning empty if the category field value is too long? How was this handled before?

Reply:
The consequence is we skip the entity and not analyze it.  The logic is the same as before.